### DISCUSSION TITLE
USD Hydra plugin for Blender v1.0.0 beta

### DISCUSSION THREAD

**Discussion by bsavery** [GPUOpen-LibrariesAndSDKs/BlenderUSDHydraAddon#90] [General]
This is an beta release of the Hydra Rendering Addon for Blender.

Our main goal is to open new workflows of using USD and MaterialX in Blender.  This allows:

* Assembling and manipulating USD data
* Referencing USD files, not just import and export so they don't have to be in Blender memory
* Importing and converting MaterialX networks.

We're very interested to hear feedback on how YOU want to use USD and MaterialX in Blender!

<hr /><em>This discussion was created from the release <a href='https://github.com/GPUOpen-LibrariesAndSDKs/BlenderUSDHydraAddon/releases/tag/v1.0.21-beta'>USD Hydra plugin for Blender v1.0.0 beta</a>.</em>

**Comment by nequine** [comment]
Hello guys!

Amazing work on the addon. I love to finally have a way to work with a proper deferred workflow inside of blender, and with nodes ! That's for me the biggest strength of this USD Hydra plugin.

What would be your current plan ? Is there a discussion with the blender foundation on adding Cycles to the list of disponible renderers ?

I would personally love to be able to bring in alembic/vdbs files and convert them on the fly to usd for assembly purpose. Also having more options when converting/exporting usds ( stage, overrides, layers, etc... ).

Again, thank you for this promising addon ! Keep up the good work !

**Reply by bsavery** [reply]
The problem is that Cycles does not yet support MaterialX nodes so we would have to include another code path to use cycles in here.  

If a user wanted to contribute that code though we wouldn't say no!
But overall, I do think this add-on should be native to Blender to enable more workflows.

**Reply by nequine** [reply]
Sadly, I have nothing of a dev, at all. I am just a casual blender user that works on animated features and that is interested in blender for its personal projects.

I personally have very little interest in materialX. While I think it is a cool things to have, its use case is, IMO, very specific. It is not a hazard that it is ILM that developed it. Not a lot of productions use multiple renderers on a single movie, and ILM are one of them. While I can see why some people would like to have it in blender, I strongly believe that blender/cycles have bigger issues to deal with, in the I/O department for example, and is a pain to integrate in existing pipeline.  
The USD Hydra plugin is providing a possible solution to the data management mess that can become blender on more complex scene. Deferred workflow is the way to go in my opinion, and the only way to simulate it right now in blender is by using .blend files, which is not very pipeline friendly and can add a lot of extra work to do.

Would it be possible to split USD Hydra plugin and the materialX in 2 different plugins ? Or maybe having materialX behaves the same way as OSL in cycles, just a check box that we can enable or not ?
Does adding hydra to the rendering process add some bottleneck/performance issues ?

**Reply by bsavery** [reply]
"Would it be possible to split USD Hydra plugin and the materialX in 2 different plugins ?"

If you wanted to make the USD Hydra plugin ONLY work with Cycles as a renderer, then yes it would be possible.  That is not our goal.  We want to make a hydra plugin that works across renderers.

Basically we have to send shader nodes to USD.  We could just send cycles nodes, but no other renderer other than cycles would know what to do with them.  MaterialX Provides a "common definition" of nodes that is meant to be portable across renderers.  

So I get what you're asking.  You want to use this file referencing to solve the blender "big data" problem.  And totally legit.  You could totally do this with the USD Hydra plugin and using the [Cycles Hydra delegate](https://github.com/tangent-opensource/hdBlackbird).  But it would need one of two things.
1.  Edit the above linked delegate to translate MaterialX -> Cycles nodes. (in all honesty probably not that hard)
2. Edit this plugin to pass straight through the cycles node if Cycles delegate is selected.  Also probably not that hard and we would accept a pull request for this but not work we're going to focus on as our first priority is generic across renderers.

Good discussion!

**Reply by nequine** [reply]
Doesn't USD has its own shader solution ?

Adding an automatic "pass by" to skip the materialX part when dealing with Cycles could be a solution indeed.

With the USD Hydra plugin, is your objectives to bring a  solution to other vendors with hydra compatible renderer to add their software to blender ? Or maybe just give the options to individual developers ?

**Reply by bsavery** [reply]
"Adding an automatic "pass by" to skip the materialX part when dealing with Cycles could be a solution indeed. (number 2 I listed above)

With the USD Hydra plugin, is your objectives to bring a solution to other vendors with hydra compatible renderer to add their software to blender ? Or maybe just give the options to individual developers ?"

Our goal is to make a generic solution for rendering in Hydra see:  [readme](https://github.com/GPUOpen-LibrariesAndSDKs/BlenderUSDHydraAddon#blender-usd-hydra-addon)

And for doing the USD compositing and referencing.  Doing a specific passthrough for Cycles shaders is fine, but not the priority.  Like I said we're happy to include it!

**Reply by nequine** [reply]
Thanks for your time and all the precisions, I appreciate it.
I've been following what the folks at AMD have been doing lately, thanks for the hard work guys! And thanks for keeping most of the things opensource ;)

Keep it up!

**Comment by marquismiller** [comment]
Very epic. But a lot of us are into the real-time as well as stylized/anime works of EEVEE. Is there a chance you'll be able to work a way to include this renderer? 

Really... REALLY do hope so. EEVEE needs it lol.

**Reply by bsavery** [reply]
Hi there.  If you take a look at my video:  https://www.youtube.com/watch?v=T7l8rVUvMbs&t=141s You'll see we also include the Hydra GL renderer which I'd say in my testing is faster than EEVEE

**Reply by marquismiller** [reply]
@bsavery That's the thing. If Cycles has a delegate, could EEVEE? If not EEVEE, could Hydra GL access files that are real-time/procedural or EEVEE exclusive into MaterialX? Examples would be Lightning Boy https://www.youtube.com/watch?v=8fHZcnTFYEI
A good chunk of Blender user only choose EEVEE for NPR/stylized goals. But NPR still need some way to optimize.
Hoping this could be the "it" factor many are missing.

**Reply by bsavery** [reply]
I have no idea what's inside their shader but it could probably be recreated in MaterialX and used with the GL renderer.  

Basically any node material that works in cycles / EEVEE could be ported to materialX

**Comment by sketchpunk** [comment]
If possible, I'd love for the ability to export blender's shader graphs to materialX.  Importing is great but going the other way around is even better.

**Reply by DagerD** [reply]
Recently we added a functionality to export Blender nodegraph to MaterialX nodegraph, you can track progress on this page https://projects.blender.org/blender/blender/issues/112864. It works with HdStorm and HdRPR render delegates. 

But there is only one option to see content of MaterialX - run Blender with these options: `--log-level 1 --log materialx.shader`. We have plans to add the ability to save Blender nodegraph to MaterialX file.

You need to download latest Blender 4.1.0 Alpha or 4.0.0 Beta for this.

**Comment by BernardoI** [comment]
Hi! This is a super interesting project! Is it still being worked on? I tried the Hydra plugin in blender 4.3 and I can't find the USD node editor. I hope the project haven't been abandoned, I think blender is in dire need of such a pipeline
